implementations/rust/ockam/ockam_api/src/nodes/service/tcp_inlets/session_replacer/mod.rs

```diff
@@ -0,0 +1,449 @@
+use ockam_transport_tcp::new_certificate_provider_cache;
+use std::sync::{Arc, Weak};
+use std::time::Duration;
+
+use tokio::time::timeout;
+
+use crate::error::ApiError;
+use crate::nodes::connection::Connection;
+use crate::nodes::service::certificate_provider::ProjectCertificateProvider;
+use crate::nodes::service::tcp_inlets::access_control::inlet_access_control;
+use crate::nodes::service::tcp_inlets::terminal_notifier::TcpInletNotifier;
+use crate::nodes::service::SecureChannelType;
+use crate::nodes::NodeManager;
+use crate::session::replacer::{
+    ActiveInletRoute, AdditionalSessionReplacer, ReplacerOutcome, ReplacerOutputKind,
+    SessionReplacer, MAX_RECOVERY_TIME,
+};
+use crate::DefaultAddress;
+use ockam::identity::{Identifier, SecureChannel};
+use ockam::tcp::TcpInletOptions;
+use ockam::udp::{UdpPuncture, UdpPunctureNegotiation, UdpTransport};
+use ockam::Result;
+use ockam_abac::{PolicyExpression, Resource};
+use ockam_core::errcode::{Kind, Origin};
+use ockam_core::{async_trait, route, Error, Route, TryClone};
+use ockam_multiaddr::MultiAddr;
+use ockam_node::Context;
+use ockam_transport_tcp::TcpInlet;
+use selector::OutletMultiAddrSelector;
+
+pub(super) mod selector;
+
+pub(super) struct InletParameters {
+    pub(super) terminal_notifier: TcpInletNotifier,
+    pub(super) outlet_address_selector: OutletMultiAddrSelector,
+    pub(super) prefix_route: Route,
+    pub(super) suffix_route: Route,
+    pub(super) authorized: Option<Identifier>,
+    pub(super) wait_for_outlet_duration: Duration,
+    pub(super) resource: Resource,
+    pub(super) policy_expression: Option<PolicyExpression>,
+    pub(super) secure_channel_identifier: Option<Identifier>,
+    pub(super) disable_tcp_fallback: bool,
+    pub(super) tls_certificate_provider: Option<MultiAddr>,
+    pub(super) enable_nagle: bool,
+    pub(super) skip_handshake: bool,
+}
+
+/// The status of the additional Inlet session.
+pub(super) struct AdditionalInletSessionStatus {
+    pub(super) secure_channel: SecureChannel,
+    pub(super) udp_puncture: UdpPuncture,
+}
+
+/// The status of the Inlet session.
+pub(super) struct InletSessionStatus {
+    pub(super) main_route: Route,
+    pub(super) connection: Connection,
+    pub(super) last_route_key: String,
+    pub(super) additional: Option<AdditionalInletSessionStatus>,
+    pub(super) original_multiaddr: MultiAddr,
+}
+
+pub(super) struct InletSessionReplacer {
+    pub(super) context: Context,
+    pub(super) node_manager: Weak<NodeManager>,
+    pub(super) inlet: Arc<TcpInlet>,
+    pub(super) udp_transport: Option<Arc<UdpTransport>>,
+    pub(super) status: Option<InletSessionStatus>,
+    pub(super) parameters: Arc<InletParameters>,
+}
+
+impl InletSessionReplacer {
+    /// Returns a standalone instance of the replacer, it shared the dependencies but not
+    /// the status
+    pub(super) fn standalone(&self) -> Result<InletSessionReplacer> {
```

Comment:
I'm not getting the meaning of standalone in this context. Using "python terms", isn't this a "shallow clone"?

Reply:
It's a partial shallow clone? :)
Would `clone_without_status()` be more readable?

Reply:
Sounds good to me 👍